Enforce bootloader present or specified as not present in ChibiOS builds

peterbarker opened the pull request:
Currently in master if a bootloader isn't present then we silently ignore that problem, leading to PRs like @Hwurzburg 's here: https://github.com/ArduPilot/ardupilot/pull/24565

Prevent that happening by either requiring the bootloader to be present or specifying that it is not to be embedded in the using a directive in the main firmware's hwdef.

Based on top of `chibios_hwdef.py` cleanups here: https://github.com/ArduPilot/ardupilot/pull/24572

Hwurzburg reviewed (commented):
RADIX2HD I think needs the define AP_BOOTLOADER_FLASHING_ENABLED 0 also

peterbarker commented:
> I like it - but I think it would have been better if you put the flak8 cleanup in a separate PR since that is the majority of the churn

I think you later realised I did do that :-)

When I do do that I try to remember to put a comment in.  Which I did do in the opening comment of this one.

... actually, I now realise it's also including a fix for CubeOrange-SimOnHardware, which is also in a separate PR...

peterbarker commented:
> RADIX2HD I think needs the define AP_BOOTLOADER_FLASHING_ENABLED 0 also

Yep.  I'm iteratively running `size_compare_branches.py` locally, adding boards as they pop up.

I'm hoping that I can eliminate the blacklist here: https://github.com/ardupilot/ardupilot/blob/master/Tools/scripts/size_compare_branches.py#L147 and ask the `chibios_hwdef.py` script whether building a bootloader is appropriate.  It's not quite the same thing as we might be *capable* of building a bootloader, just not want to...

peterbarker commented:
I've gone through and added the relevant define everywhere where it's required.

I'm *quite* sure we should be adding more bootloaders to the `Tools/bootloaders` directory based of the results here!

I've run this through size_compare_branches and the only changes are to boards where the new define has been added to teh hwdef in romfs.

```
pbarker@fx:~/rc/ardupilot(pr/must-have-bootloader)$ git show | fgrep -- '--- a/libraries'
--- a/libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4Pilot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/ARKV6X/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/C-RTK2-HP/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/CarbonixF405/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/CubeBlack-periph/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/CubeOrange-SimOnHardWare/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/CubeOrangePlus-SimOnHardWare/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/G4-ESC/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL_intf/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/HerePro/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/MatekH743-periph/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/MatekL431-HWTelem/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/MatekL431-Proximity/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/MatekL431-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/MazzyStarDrone/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Nucleo-L476/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Nucleo-L496/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/NucleoH755/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/PH4-mini-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/PixFlamingo/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/PixSurveyA1/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Pixhawk1-1M-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Pixhawk1-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Pixhawk6X-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Sierra-F405/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Sierra-F412/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/Sierra-F9P/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/YJUAV_A6/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/f303-PWM/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/f303-TempSensor/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/fmuv2/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/fmuv3-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/kha_eth/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/luminousbee4/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/mRoControlZeroClassic/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/mRoControlZeroH7-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/mRoControlZeroH7/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/omnibusf4pro-one/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/revo-mini-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/revo-mini-i2c-bdshot/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/revo-mini-i2c/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/skyviper-f412-rev1/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/skyviper-journey/hwdef.dat
--- a/libraries/AP_HAL_ChibiOS/hwdef/skyviper-v2450/hwdef.dat
pbarker@fx:~/rc/ardupilot(pr/must-have-bootloader)$ 
```

I think we should also consider adding a hwdef option to say, "use the bootloader from this other board over here" so on derivative boards where we don't change the board ID or USB strings or whatever then we don't need a separate board ID.

Hwurzburg reviewed (changes requested):
I agree that we should enforce this, but disagree that we should skip fixing the boards that DO need bootloaders by just tagging them as NOT requiring the bootloader

peterbarker commented:
> I agree that we should enforce this, but disagree that we should skip fixing the boards that DO need bootloaders by just tagging them as NOT requiring the bootloader

I have absolutely zero interest in adding these myself.  If you want to push more patches into a separate PR based on this one, please do.  It is, I agree, something that needs fixing.  This PR just stops things getting worse.

That being said, if we can get consensus on the "include the bootloader from this other board" being a good thing, I will implement that modification to the infrastructure.

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/mRoControlZeroH7-bdshot/hwdef.dat:
```diff
@@ -14,3 +14,6 @@ PD13 TIM4_CH2 TIM4 PWM(5) GPIO(54) BIDIR
 PI6  TIM8_CH2 TIM8 PWM(8) GPIO(57)
+
+undef AP_BOOTLOADER_FLASHING_ENABLED
+define AP_BOOTLOADER_FLASHING_ENABLED 1
```
is this needed?

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/C-RTK2-HP/hwdef.dat:
```diff
@@ -121,3 +121,6 @@ define AP_GPS_NMEA_CUSTOM_CONFIG_STRING "CONFIG COM1 230400 8 N 1"
 define NMEA_UNICORE_SETUP "CONFIG COM1 230400 8 n 1\r\nGPGGA 0.2\r\n"
+
+# bootloader embedding / bootloader flashing not available
+define AP_BOOTLOADER_FLASHING_ENABLED 0
```
add bl for this one

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/CarbonixF405/hwdef.dat:
```diff
@@ -139,7 +139,5 @@ define HAL_PERIPH_ADSB_BAUD_DEFAULT 57600
-
-
-
+# bootloader embedding / bootloader flashing not available
```
add bl for carbonix boards

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat:
```diff
@@ -142,7 +142,5 @@ define HAL_PERIPH_ADSB_BAUD_DEFAULT 57600
-
-
-
+# bootloader embedding / bootloader flashing not available
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/G4-ESC/hwdef.dat:
```diff
@@ -153,3 +153,6 @@ define CAN_APP_NODE_NAME "org.ardupilot.G4-ESC"
 # don't share any DMA channels (there are enough for everyone)
 DMA_NOSHARE *
+
+# bootloader embedding / bootloader flashing not available
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MatekL431-Proximity/hwdef.dat:
```diff
@@ -11,3 +11,5 @@ define HAL_PROXIMITY_ENABLED 1
 
 define AP_PERIPH_PRX_PORT_DEFAULT 2
 
+# bootloader embedding / bootloader flashing not available
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MatekL431-bdshot/hwdef.dat:
```diff
@@ -33,3 +33,6 @@ define HAL_PERIPH_ENABLE_NOTIFY
 define HAL_WITH_ESC_TELEM 1
+
+# bootloader embedding / bootloader flashing not available
+define AP_BOOTLOADER_FLASHING_ENABLED 0
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MazzyStarDrone/hwdef.dat:
```diff
@@ -163,3 +163,6 @@ define STM32_PWM_USE_ADVANCED TRUE
 # we are low on flash on this board
 include ../include/minimize_features.inc
+
+# bootloader embedding / bootloader flashing not available
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/PH4-mini-bdshot/hwdef.dat:
```diff
@@ -14,3 +14,6 @@ PA10 TIM1_CH3 TIM1 PWM(2) GPIO(51) BIDIR
 DMA_PRIORITY TIM1_CH3 TIM1_CH2 TIM1_UP SDMMC* USART2* ADC* SPI* TIM*
+
+# bootloader embedding / bootloader flashing not available
+define AP_BOOTLOADER_FLASHING_ENABLED 0
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/PixSurveyA1/hwdef.dat:
```diff
@@ -19,3 +19,6 @@ define HAL_IMU_TEMP_DEFAULT 45
 define HAL_IMU_TEMP_MARGIN_LOW_DEFAULT 5
+
+# bootloader embedding / bootloader flashing not available
+define AP_BOOTLOADER_FLASHING_ENABLED 0
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/Pixhawk1-bdshot/hwdef.dat:
```diff
@@ -36,3 +36,6 @@ define HAL_PROBE_EXTERNAL_I2C_COMPASSES
 define BOARD_CHECK_F427_USE_1M "ERROR: 1M flash use Pixhawk1-1M"
+
+# bootloader embedding / bootloader flashing not available
+define AP_BOOTLOADER_FLASHING_ENABLED 0
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/Pixhawk6X-bdshot/hwdef.dat:
```diff
@@ -14,3 +14,6 @@ PD13 TIM4_CH2 TIM4 PWM(5) GPIO(54) BIDIR
 
 DMA_PRIORITY TIM5* TIM4* SPI1* SPI2* SPI3* SDMMC* USART6* ADC* UART* USART*
+
+# bootloader embedding / bootloader flashing not available
```
add bl

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/f303-PWM/hwdef.dat:
```diff
@@ -30,3 +30,5 @@ PB15 TIM1_CH3N TIM1 PWM(4) GPIO(53)
 # enable PWM
 define HAL_PERIPH_ENABLE_RC_OUT
 
+# bootloader embedding / bootloader flashing not available
```
add bl

tridge reviewed (changes requested):
tagged ones that are easy decision to add bootloader

peterbarker commented:
I've rebased on top of Henry's PR which adds bootloaders for the ones tridge wanted here.

I removed the patch which added the exceptions and ran configure-all, and then added back in the ones required.  configure-all passed after that patch was added.

I'll merge this tomorrow after looking through it again....
